libs/metagen/src/client_py/types.rs

```diff
@@ -0,0 +1,298 @@
+// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
+// SPDX-License-Identifier: MPL-2.0
+
+use std::fmt::Write;
+
+use common::typegraph::*;
+
+use super::utils::{normalize_struct_prop_name, normalize_type_title};
+use crate::{interlude::*, shared::types::*};
+
+pub struct PyTypeRenderer {}
+impl PyTypeRenderer {
+    fn render_alias(
+        &self,
+        dest: &mut impl Write,
+        alias_name: &str,
+        aliased_ty: &str,
+    ) -> std::fmt::Result {
+        writeln!(dest, "{alias_name} = {aliased_ty}")?;
+        writeln!(dest)?;
+        Ok(())
+    }
+
+    /// `props` is a map of prop_name -> (TypeName, serialization_name)
+    fn render_object_type(
+        &self,
+        dest: &mut impl Write,
+        ty_name: &str,
+        props: IndexMap<String, (Rc<str>, bool)>,
+    ) -> std::fmt::Result {
+        writeln!(dest, r#"{ty_name} = typing.TypedDict("{ty_name}", {{"#)?;
+        for (name, (ty_name, _optional)) in props.into_iter() {
+            // FIXME: use NotRequired when bumping to python version
+            // that supports it
+            // also, remove the total param below
+            // if optional {
```

Comment:
Wouldn't this depend on the user's python version?
But I guess we can always check the available python before generating..

Reply:
Python is backward compatible so we can stay on 3.8 perpetually I suppose but since 3.8 is EOL soon, we can just track whichever earliest python version has LTS support upstream.

Reply:
👍